server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -257,7 +260,8 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
         List<String> pathUnmatch = new ArrayList<>(4);
         Map<String, Object> mapping = null;
         boolean runtime = false;
-        String matchMappingType = null;
+        Object matchMappingType = null;
+        Object unmatchMappingType = null;
```

Comment:
Let's use a `List<String>` here that only has one element if the `match_mapping_type`/`unmatch_mapping_type` was provided as a string.

Reply:
Done